packages/fdc3-agent-proxy/src/messaging/AbstractMessaging.ts

```diff
@@ -30,48 +29,59 @@ export abstract class AbstractMessaging implements Messaging {
     }
 
     getSource(): AppIdentifier {
-        return this.appId!!
+        if(this.appId === null){
+            throw new Error("Unconnected. No source available.")
+        }
+        return this.appId
     }
 
     waitFor<X>(filter: (m: any) => boolean, timeoutErrorMessage?: string): Promise<X> {
         const id = this.createUUID()
         return new Promise<X>((resolve, reject) => {
-            var done = false;
-            const l: RegisterableListener = {
+            let done = false;
+            const listener: RegisterableListener = {
                 id,
                 filter: filter,
                 action: (m) => {
                     done = true
                     this.unregister(id)
                     resolve(m)
+                },
+                register: async () => {
+                    this.register(listener)
+                },
+                unsubscribe: async () => {
+                    this.unregister(id);
                 }
-            } as RegisterableListener
-
+            }
 
-            this.register(l);
+            listener.register();
 
-            if (timeoutErrorMessage) {
-                setTimeout(() => {
-                    this.unregister(id)
-                    if (!done) {
-                        console.error(`Rejecting after ${this.timeout}ms with ${timeoutErrorMessage}`)
-                        reject(new Error(timeoutErrorMessage))
-                    }
-                }, this.timeout);
+            if (!timeoutErrorMessage) {
+                return;
             }
 
+            setTimeout(() => {
+                this.unregister(id)
+                if (!done) {
+                    console.error(`Timed out after ${this.timeout}ms with ${timeoutErrorMessage}`)
+                    reject(new Error(timeoutErrorMessage))
+                }
+            }, this.timeout);
         })
     }
 
 
     async exchange<X>(message: any, expectedTypeName: string, timeoutErrorMessage?: string): Promise<X> {
         const errorMessage = timeoutErrorMessage ?? `Timeout waiting for ${expectedTypeName} with requestUuid ${message.meta.requestUuid}`
-        const prom = this.waitFor(m =>
+        const messageResponseWaiter = this.waitFor<X>(m =>
             (m.type == expectedTypeName)
             && (m.meta.requestUuid == message.meta.requestUuid), errorMessage)
         this.post(message)
-        const out: any = await prom
+        const out = await messageResponseWaiter
+        // @ts-expect-error: TODO: We should make a list of all of the possible messages that can be called through this, and make that list the generic. This will ensure better type safety.
```

Comment:
I'm pretty sure that relevant types already exist for that: AppRequestMessage, AgentResponseMessage, AgentEventMessage were intended for such usage and contain unions of the relvant type strings for messages in their set.

Reply:
Those types can't be easily applied yet due to these functions serving both WCP and DACP - I'll fix this in my branch where I've unpicked that already